### PR TITLE
feat(slider): add section labels support

### PR DESCRIPTION
**Pull request by scharinger**
### Describe your changes

We need a slider with sections that have labels and customable track colors.

### Checklist before requesting a review

- [x] I have performed a self-review of my own code
- [x] I have verified that the code builds perfectly fine on my local system
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have verified that my code follows the style already available in the repository
- [x] I have made corresponding changes to the documentation

image

### CODE REVIEW

File: components/slider/src/slider.types.ts
Comment (trew) [inline components/slider/src/slider.types.ts:47]: It doesn't have to be boolean. Marks are typed as boolean since if you set "marks": true together with "steps", you will get tiny marks on the track that highlights the step values. See the "With steps" example.

File: components/slider/src/use-range-slider.ts
Comment (trew) [inline components/slider/src/use-range-slider.ts:396]: ?

File: components/slider/src/use-slider-styles.ts
Comment (trew) [inline components/slider/src/use-slider-styles.ts:225]: Conditional memos is a no no. Not sure why there isn't a lint warning or something for this, but this needs to be rewritten to have the memo run always.

File: examples/src/components/story/story-page.tsx
Comment (trew) [inline examples/src/components/story/story-page.tsx:27]: needed?

File: examples/src/components/navigation-menu/navigation-menu.tsx
Comment (trew) [inline examples/src/components/navigation-menu/navigation-menu.tsx:34]: needed?

**Review by trew — CHANGES_REQUESTED** [review]
Overall it looks nice. Some changes required though. I defer to other maintainers whether or not changes such as changeset.js/ts should be included or not. They are good changes but have nothing to do with the PR.

File: biome.json
Comment (scharinger) [reply biome.json:1]: Because the developer might have it installed on their workstation/_vscode setups_ for other projects. And if they are like me, they don't know how to turn it off for specific workspaces and get biome's linter checking the source code of every source code, like this one and make the development experience harder. So it's not a need, more making life easier for developers like me. :smiley_cat:

Comment (axis-d0op) [reply biome.json:1]: We would prefer not to involve another lint:ing framework.
if it feels essential it then should be be part of own PR.

Comment (scharinger) [reply biome.json:1]: I can leave it out of this PR, but I would still argue that we are not involving another linting framework as long as we don’t force it in any way or use it in the action runners for linting. By adding the suggested `biome.js` file, we would merely be helping contributors who have Biome installed by preventing disruptions similar to those caused by the disabled Biome lint rule.

File: components/slider/src/render-slider.tsx
Comment (scharinger) [reply components/slider/src/render-slider.tsx:19]: Yes, I found out the issue you are mentioning. There was a issue I had that I solved this sub-optimal way. I have to try to find a solution on either the first or the new issue that you also found.

File: components/slider/src/section/label/section-label.types.ts
Comment (scharinger) [reply components/slider/src/section/label/section-label.types.ts:19]: Maybe `from` and `to`?

Comment (scharinger) [reply components/slider/src/section/label/section-label.types.ts:19]: Is there an easy way to test the components in `rlt` mode?

File: components/slider/src/section/label/section-label.tsx
Comment (scharinger) [reply components/slider/src/section/label/section-label.tsx:9]: Very true. I started with the label and later started implementing the track section, thus the naming :slightly_smiling_face:.

File: components/slider/src/slider.types.ts
Comment (scharinger) [reply components/slider/src/slider.types.ts:47]: Aha, thank you for the explanation!

File: biome.json
Comment (scharinger) [reply biome.json:1]: I just found out that a vs code extension can be disabled for the current workspace. That's another way to solve my issue without introducing the `biome.json` file.

File: components/slider/src/mark/label/use-mark-label.ts
Comment (scharinger) [reply components/slider/src/mark/label/use-mark-label.ts:18]: Updated.

File: components/slider/src/render-slider.tsx
Comment (scharinger) [reply components/slider/src/render-slider.tsx:19]: The issue I was solving was to not crop the marks, but that can be achieved by only moving out the marks from the rail but keep the track inside. Which I do now. So far I still have issues though when min is not `0` with the marks and sections and it's labels.

Marks that where cropped before.
image

File: components/slider/src/use-range-slider.ts
Comment (scharinger) [reply components/slider/src/use-range-slider.ts:396]: I can remove the comment. It just helps me remembering that `minValue` contains from my way of seeing it the start if the range and not the minimum possible or allowed value.

File: components/slider/src/use-slider-styles.ts
Comment (scharinger) [reply components/slider/src/use-slider-styles.ts:225]: Yeah, I'm used to memos to not be allowed in conditions and also noticed the linter not to catch it. I figured the consolidation outcome can not be changed as long as the range slider state is not changed but I can try to find a better way, and allowed way to di it.

File: components/slider/src/use-range-slider.ts
Comment (scharinger) [reply components/slider/src/use-range-slider.ts:396]: Here `minValue` is `selectedMinValue` but could also be falsely read as `allowedMinValue`. Therefore my commend. But maybe we could improve the variable names instead someday.

File: examples/src/components/navigation-menu/navigation-menu.tsx
Comment (scharinger) [reply examples/src/components/navigation-menu/navigation-menu.tsx:34]: No. It helped me test the slider with our use cases and I saw no reason why it would need to be wider. I can revert it.

File: examples/src/components/story/story-page.tsx
Comment (scharinger) [reply examples/src/components/story/story-page.tsx:27]: No. Reverted. But a lot of screen space is wasted IMO.

File: components/slider/src/section/label/section-label.types.ts
Comment (trew) [reply components/slider/src/section/label/section-label.types.ts]: In the examples project, go to Themes and to the right there is a Swap direction toggle.In the examples project, go to Themes and to the right there is a Swap direction toggle.

**Comment by trew** [comment]
The "Sweet spot" mark is  example is on 45.5, but the steps of the slider only allow 45 and 46. However, if I click the "Sweet spot" label, exactly 45.5 is selected. What do we want as an intended behavior? I'm thinking snapping to the closest allowed value.

**Comment by trew** [comment]
The range slider doesn't seems to work that well with sections. The track goes out of bounds. Also the mark label highlighting doesn't work properly when multiple marks have values that equal them.

File: components/slider/src/section/label/section-label.types.ts
Comment (scharinger) [reply components/slider/src/section/label/section-label.types.ts]: Great! My changes seems to work in rtl too. 🙂
image

**Comment by scharinger** [comment]
> The range slider doesn't seems to work that well with sections. The track goes out of bounds. Also the mark label highlighting doesn't work properly when multiple marks have values that equal them.

Updated the gradient color logic to work for the range slider also.

image

**Comment by trew** [comment]
> Only problem is that the thumb color on range sliders act oddly. ... We currently don't have a use case for range sliders using sections so maybe it could wait until we need it?

Yes. That one is ok for me to skip with this PR.

One final thing I noticed behavior-wise is that the active state for mark labels (when activeEqual is true) only works for the max value. That should probably test for all values. 

Only "Boiling" is active. "Freezing" should be as well.
image

**Comment by scharinger** [comment]
> > Only problem is that the thumb color on range sliders act oddly. ... We currently don't have a use case for range sliders using sections so maybe it could wait until we need it?
> 
> Yes. That one is ok for me to skip with this PR.
> 
> One final thing I noticed behavior-wise is that the active state for mark labels (when activeEqual is true) only works for the max value. That should probably test for all values.
> 
> Only "Boiling" is active. "Freezing" should be as well. image

Updated the that the lower range value also is set as `active`.

image

File: components/slider/src/mark/label/use-mark-label.ts
Comment (trew) [inline components/slider/src/mark/label/use-mark-label.ts:18]: This needs to check all possible values, since sliders with 3 or more values are also supported.

```suggestion
    ? values.some((value) => value === props.value)
```

File: examples/src/stories/slider/examples/range-slider-with-section-example.tsx
Comment (trew) [inline examples/src/stories/slider/examples/range-slider-with-section-example.tsx:31]: Do something with this. Remove or mirror to the string section below.

Comment (scharinger) [reply examples/src/stories/slider/examples/range-slider-with-section-example.tsx:31]: Removed.

File: components/slider/src/mark/label/use-mark-label.ts
Comment (scharinger) [reply components/slider/src/mark/label/use-mark-label.ts:18]: Aha, did not think about more than two. Updated the gradient logic to work with 3 or more.
image

**Review by trew — APPROVED** [review]

**Comment by trew** [comment]
@axis-d0op The unresolved conversations blocks merging. How do I resolve them?

**Comment by scharinger** [comment]
Rebased.

**Review by axis-d0op — APPROVED** [review]
LGTM / samuel & jimmy

**Review by trew — APPROVED** [review]

**Pull request merged**
